### PR TITLE
Duplicate product/service line (auto fill input)

### PR DESCRIPTION
**Pull request by andrelec1**
Button for duplication added in bills editor
this button fulfill the values for the new line

### CODE REVIEW

**Comment by gonzalesleo** [comment]
Very useful for us, and many customers we have would benefit from it.

File: htdocs/core/tpl/objectline_view.tpl.php [hunk 1]
```diff
@@ -148,6 +149,32 @@
...
+				$string_argument .= ", '".base64_encode($line->description)."'";
+				$string_argument .= ", '".$line->qty."'";
+				$string_argument .= ", '".$line->remise_percent."'";
+				$string_argument .= ", '".$line->subprice."'";
+				$string_argument .= ", '".$line->tva_tx."'";
+				$string_argument .= ", '".$line->product_type."'";
+			 }	?>
+			<td align="center"><?php $coldisplay++; ?>
+				<a href="" class="dupdup" onclick="autofillnewproduct(<?php echo $string_argument; ?>); return false;" >
+					<?php echo img_picto("Dupliquer",'split.png'); ?>
```
Comment (jfefe) [inline htdocs/core/tpl/objectline_view.tpl.php]: Missing translation here. Please fix it, thanks.

File: htdocs/admin/facture.php [hunk 2]
```diff
@@ -261,6 +262,24 @@
     }
 }
 
+if ($action == 'setaffdupliquer')
+{
+	$affdupliquer = GETPOST('affdupliquer','alpha');
+
+	$res = dolibarr_set_const($db, "FAC_AFF_BUTTON_DUPPLIQUER",$affdupliquer,'chaine',0,'',$conf->entity);
```
Comment (jfefe) [inline htdocs/admin/facture.php]: Please use a standard name...
Why not INVOICE_SHOW_DUPLICATE_BUTTON (or INVOICE_SHOW_DUPLICATE_LINE_BUTTON or if it's general use SHOW_DUPLICATE_LINE_BUTTON ?

Comment (andrelec1) [inline htdocs/admin/facture.php]: i juste copy/past the "FAC_FORCE_DATE_VALIDATION"  ...
but i can change this.

Comment (jfefe) [inline htdocs/admin/facture.php]: Yep but we must keep internationalisation in mind for the future ;)

**Comment by eldy** [comment]
To encode sql request content, you must use  db->escape and not base64_encode function.

**Comment by andrelec1** [comment]
I don't use base64_encode for encode sql request ...
I use this function for send a string generate in php at the JavaScript ( and i have some probleme with multiple special char ... )

**Comment by andrelec1** [comment]
No news ?

File: htdocs/compta/facture.php [hunk 3]
```diff
@@ -3325,6 +3326,32 @@
...
+		print '				$("#dp_desc").val(atob(arguments[1]));';
+		print '				$("#qty").val(Number(arguments[2]));';
+		print '				$("#remise_percent").val(Number(arguments[3]));';
+		print '				$("#price_ht").val(Number(arguments[4]));';
+		print '				$("#tva_tx").val(Number(arguments[5]));';
+		print '				$("#select_type").val(Number(arguments[6]));';
+		print '     	}';
+		print '         else if(arguments[0] == 2){';
+		print '             $("#prod_entry_mode_predef").click();';
+		print '				$("#dp_desc").val(atob(arguments[1]));';
```
Comment (gonzalesleo) [inline htdocs/compta/facture.php:3344]: The base64 is decoded here with "atob"
(see below for the encode)

File: htdocs/core/tpl/objectline_view.tpl.php [hunk 4]
```diff
@@ -148,6 +149,32 @@
 		<?php } ?>
 	</td>
 
+	<?php if(!empty($conf->global->INVOICE_SHOW_DUPLICATE_BUTTON )){
+			if(!empty($line->ref)){
+				//mode predef
+				$string_argument = "'2'";
+				$descclean = str_replace($line->product_desc ."\n", "",$line->description);
+				$string_argument .= ", '".base64_encode($descclean)."'";
```
Comment (gonzalesleo) [inline htdocs/core/tpl/objectline_view.tpl.php:157]: The base64 is encoded here.
Nothing to do with SQL :)

**Comment by eldy** [comment]
That's the point to encode data, you must use escape function according to context (db->escape for sql forging, dol_escape_html for html tag, dol_escape_js for js script...)

**Pull request closed without merging**